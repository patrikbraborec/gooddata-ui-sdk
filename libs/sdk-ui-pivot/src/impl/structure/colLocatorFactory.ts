// (C) 2021 GoodData Corporation

import { DataColGroup, DataColLeaf, isDataColGroup } from "./tableDescriptorTypes";
import { ColumnLocator, IAttributeColumnLocator, IMeasureColumnLocator } from "../../columnWidths";
import { IAttributeDescriptor, IMeasureDescriptor, IResultAttributeHeader } from "@gooddata/sdk-backend-spi";
import { invariant } from "ts-invariant";
import zip from "lodash/zip";

function createAttributeLocator(
    descriptor: IAttributeDescriptor | undefined,
    header: IResultAttributeHeader | undefined,
): IAttributeColumnLocator {
    // if this bombs it means something is wrong with the col descriptors or in very bad case in the DVF data access logic.
    // by contract, all data series have same number of descriptors & headers. therefore the zipping logic should never
    // run into situation where the arrays are of different size.
    invariant(descriptor && header);

    return {
        attributeLocatorItem: {
            attributeIdentifier: descriptor.attributeHeader.localIdentifier,
            element: header.attributeHeaderItem.uri,
        },
    };
}

function createMeasureLocator(descriptor: IMeasureDescriptor): IMeasureColumnLocator {
    return {
        measureLocatorItem: {
            measureIdentifier: descriptor.measureHeaderItem.localIdentifier,
        },
    };
}

/**
 * Given either data col group or a data col leaf, this function will create column locator that can be used in
 * width items.
 *
 * @param col - col definition to get locators for
 */
export function createColumnLocator(col: DataColGroup | DataColLeaf): ColumnLocator[] {
    if (isDataColGroup(col)) {
        const { descriptorsToHere, headersToHere } = col;
        const descriptorsAndHeaders = zip(descriptorsToHere, headersToHere);
        descriptorsAndHeaders.push([col.attributeDescriptor, col.header]);

        return descriptorsAndHeaders.map(([descriptor, header]) =>
            createAttributeLocator(descriptor, header),
        );
    } else {
        const result: ColumnLocator[] = [];

        if (col.seriesDescriptor.attributeDescriptors && col.seriesDescriptor.attributeHeaders) {
            const descriptorAndHeaders = zip(
                col.seriesDescriptor.attributeDescriptors,
                col.seriesDescriptor.attributeHeaders,
            );

            descriptorAndHeaders.forEach(([descriptor, header]) =>
                result.push(createAttributeLocator(descriptor, header)),
            );
        }

        result.push(createMeasureLocator(col.seriesDescriptor.measureDescriptor));

        return result;
    }
}
