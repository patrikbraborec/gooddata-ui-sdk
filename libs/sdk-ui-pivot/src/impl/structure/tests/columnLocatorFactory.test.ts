// (C) 2007-2021 GoodData Corporation

import {
    SingleMeasureWithRowAttribute,
    SingleMeasureWithTwoRowAndTwoColumnAttributes,
} from "./table.fixture";
import { TableDescriptor } from "../tableDescriptor";
import { createColumnLocator } from "../colLocatorFactory";
import { DataColGroup } from "../tableDescriptorTypes";

describe("createColumnLocator", () => {
    it("creates valid leaf column locator in table without column attributes", () => {
        const t = TableDescriptor.for(SingleMeasureWithRowAttribute);

        expect(createColumnLocator(t.headers.leafDataCols[0])).toMatchSnapshot();
    });

    it("creates valid leaf column locator in table with column attributes", () => {
        const t = TableDescriptor.for(SingleMeasureWithTwoRowAndTwoColumnAttributes);

        expect(createColumnLocator(t.headers.leafDataCols[0])).toMatchSnapshot();
    });

    it("creates valid group column locator for top-level group in table with column attributes", () => {
        const t = TableDescriptor.for(SingleMeasureWithTwoRowAndTwoColumnAttributes);

        const topLevelGroup = t.headers.rootDataCols[0].children[0] as DataColGroup;

        expect(createColumnLocator(topLevelGroup)).toMatchSnapshot();
    });

    it("creates valid group column locator for second-level group in table with column attributes", () => {
        const t = TableDescriptor.for(SingleMeasureWithTwoRowAndTwoColumnAttributes);

        const secondLevelGroup = t.headers.rootDataCols[0].children[0].children[0] as DataColGroup;

        expect(createColumnLocator(secondLevelGroup)).toMatchSnapshot();
    });
});
