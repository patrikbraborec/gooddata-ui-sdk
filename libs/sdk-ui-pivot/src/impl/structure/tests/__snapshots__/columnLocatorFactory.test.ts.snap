// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createColumnLocator creates valid group column locator for second-level group in table with column attributes 1`] = `
Array [
  Object {
    "attributeLocatorItem": Object {
      "attributeIdentifier": "a_label.stage.name.stagename",
      "element": "/gdc/md/referenceworkspace/obj/1091/elements?id=166498",
    },
  },
  Object {
    "attributeLocatorItem": Object {
      "attributeIdentifier": "a_label.owner.region",
      "element": "/gdc/md/referenceworkspace/obj/1086/elements?id=460488",
    },
  },
]
`;

exports[`createColumnLocator creates valid group column locator for top-level group in table with column attributes 1`] = `
Array [
  Object {
    "attributeLocatorItem": Object {
      "attributeIdentifier": "a_label.stage.name.stagename",
      "element": "/gdc/md/referenceworkspace/obj/1091/elements?id=166498",
    },
  },
]
`;

exports[`createColumnLocator creates valid leaf column locator in table with column attributes 1`] = `
Array [
  Object {
    "attributeLocatorItem": Object {
      "attributeIdentifier": "a_label.stage.name.stagename",
      "element": "/gdc/md/referenceworkspace/obj/1091/elements?id=166498",
    },
  },
  Object {
    "attributeLocatorItem": Object {
      "attributeIdentifier": "a_label.owner.region",
      "element": "/gdc/md/referenceworkspace/obj/1086/elements?id=460488",
    },
  },
  Object {
    "measureLocatorItem": Object {
      "measureIdentifier": "m_aangOxLSeztu",
    },
  },
]
`;

exports[`createColumnLocator creates valid leaf column locator in table without column attributes 1`] = `
Array [
  Object {
    "measureLocatorItem": Object {
      "measureIdentifier": "m_aangOxLSeztu",
    },
  },
]
`;
